internal/sync/lifecycle.go

```diff
@@ -160,33 +150,29 @@ func (l *lifecycleOnce) WhenRunning(ctx context.Context) error {
 // If Stop is called multiple times it will return the error
 // from the first time it was called.
 func (l *lifecycleOnce) Stop(f func() error) error {
-	if l.stopping.Swap(true) {
-		<-l.stopCh
+	if l.state.CAS(int32(Idle), int32(Stopped)) {
+		close(l.startCh)
+		close(l.stopCh)
 		return l.err
 	}
 
-	if !l.starting.Swap(true) {
-		// Was not already starting:
-		// Pre-empt start
-		l.started.Store(true)
-		close(l.startCh)
-	} else if !l.started.Load() {
-		// Starting, but not yet started:
-		// Wait for concurrent start to finish
-		<-l.startCh
-	}
+	<-l.startCh
 
-	if l.err != nil {
-		return l.err
-	}
+	if l.state.CAS(int32(Running), int32(Stopping)) {
+		if f != nil {
+			l.err = f()
+		}
 
-	if f != nil {
-		l.err = f()
+		if l.err != nil {
+			l.state.Store(int32(Errored))
+		} else {
+			l.state.Store(int32(Stopped))
+		}
+		close(l.stopCh)
+		return l.err
 	}
-	l.errored.Store(l.err != nil)
-	l.stopped.Store(true)
-	close(l.stopCh)
 
+	<-l.stopCh
```

Comment:
Symmetrically to start(), this wait could be moved to the else clause of the previous condition.